Remove deprecated usage of ContainerAwareCommand

ValentinMerlet opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 3.x is for everything backwards compatible, like patches, features and deprecation notices
    - 4.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataPageBundle/blob/3.x/CONTRIBUTING.md#base-branch
-->
I am targeting this branch, because `4.x` will be compatible with Symfony 5.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Closes https://github.com/sonata-project/SonataPageBundle/issues/1420

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataPageBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Removed

- Removed usage of deprecated `ContainerAwareCommand` class
```

<!--
    If this is a work in progress, uncomment the "To do" section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
-->
<!--
## To do

- [ ] Update the tests;
- [ ] Update the documentation;
- [ ] Add an upgrade note.
-->

ValentinMerlet committed:
Remove deprecated usage of ContainerAwareCommand

ValentinMerlet commented on src/Command/RenderBlockCommand.php:
```diff
@@ -78,13 +125,7 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 
         // fake request
-        $request = new Request();
-        $this->getContainer()->enterScope('request');
-        $this->getContainer()->set('request', $request, 'request');
-
-        $output->writeln($this->getContainer()->get('sonata.block.renderer')->render($context));
-
-        $this->getContainer()->leaveScope('request');
```
`enterScope`/`leaveScope` functions do not exist anymore, I don't know how to reproduce this behaviour and I'm not sure what the goal was.

VincentLanglet replied:
I have no idea ^^

jordisala1991 replied:
This is for when request was an object on the container, We should ensure no one tries to access request directly from the container and use requestStack instead.

ValentinMerlet replied:
[Scopes are deprecated since Symfony 2.8](https://symfony.com/blog/new-in-symfony-2-8-deprecating-scopes-and-introducing-shared-services), is there anything to add here to keep the same behaviour as before?

jordisala1991 replied:
I would say not

VincentLanglet commented on src/Command/BaseCommand.php:
```diff
@@ -16,75 +16,112 @@
     /**
      * @return SiteManagerInterface
      */
     public function getSiteManager()
```
I don't see a reason to keep those public function when we can access by the property ?

VincentLanglet commented on src/Command/RenderBlockCommand.php:
```diff
@@ -41,21 +87,22 @@ public function configure(): void
-        if (!$manager instanceof CmsManagerInterface) {
-            throw new \RuntimeException('The service does not implement the CmsManagerInterface');
+        if (!\in_array($manager, ['sonata.page.cms.snapshot', 'sonata.page.cms.page'], true)) {
+            throw new \RuntimeException(
+                'Available managers are "sonata.page.cms.snapshot" and "sonata.page.cms.page"'
+            );
         }
 
-        $page = $manager->getPageById($input->getArgument('page_id'));
+        $managerService = 'sonata.page.cms.snapshot' === $manager ? $this->cmsSnapshotManager : $this->cmsPageManager;
```
What if someone created his own CmsManagerInterface ?
It cannot use it anymore. We should find another solution.

Maybe injecting a service locator with all the CmsManagerInterface instead of directly injecting the CmsManagerInterface.

ValentinMerlet replied:
You're right, I did not think about it. I'm trying to inject a service locator containing all the services implementing `CmsManagerInterface` as you suggest and I let you know.

ValentinMerlet replied:
Hello @VincentLanglet, I tried, as you suggest, to add a service locator for cms manager services but I'm struggling while trying to add services as argument of the service locator in a compiler pass. See the following commit: https://github.com/sonata-project/SonataPageBundle/pull/1422/commits/59edbd5b067a2694565f2d14d7ce64098e8f83ca.

I have the following error:
![error_during_compiler_pass_execution](https://user-images.githubusercontent.com/2642302/173199111-6cf34600-9284-4a94-ad61-7b9a58739a47.png). I guess using reflection in a compiler pass is not a good idea, any other idea how to add services implementing [this interface](https://github.com/sonata-project/SonataPageBundle/blob/f56de820e5d14c008238dbf7bcb077a4c963f717/src/CmsManager/CmsManagerInterface.php), maybe using tags? But it should be documented somewhere. What do you think?

VincentLanglet replied:
I know @franmomu already used service locator multiple times, maybe he can help.

If you look at https://github.com/sonata-project/SonataAdminBundle/pull/7056/files, reflection was used so this might not be the origin of the problem.

I think the purpose of the compilerPass is to dynamically call `$definition->replaceArgument` in order to set the dependency. I don't think this is what you're doing here.

ValentinMerlet replied:
> I think the purpose of the compilerPass is to dynamically call `$definition->replaceArgument` in order to set the dependency. I don't think this is what you're doing here.

I think I should use `addArgument()` as I did and not `replaceArgument()` because I guess we want to support both "natives" services that implements the [CmsManagerInterface](https://github.com/sonata-project/SonataPageBundle/blob/f56de820e5d14c008238dbf7bcb077a4c963f717/src/CmsManager/CmsManagerInterface.php) and services that implements the same interface but defined by the users of SonataPageBundle.

I wonder if the fact that I don't loop on services with a specific tag but onto [all the services](https://github.com/sonata-project/SonataPageBundle/commit/59edbd5b067a2694565f2d14d7ce64098e8f83ca#diff-91f89bb87705458d4026c45c68356cda2fd39acdb1b7cd8d73a6d94bc94b6038R28) can cause performance issues :thinking: 

**Edit**: I figured out that my issue with the fact that `Psr\SimpleCache\CacheInterface` does not exist comes from that the dependency is not installed in the vendor but I can't figure out why. Do you have an idea?

franmomu replied:
I don't know this bundle well, but reading the comments we could maybe autoconfigure services that implement `CmsManagerInterface` with a tag, that would allow to loop them instead of all services.

VincentLanglet reviewed (commented):
There are a lot of dependency in those commands ; any idea how to reduce this @jordisala1991  @sonata-project/contributors ?

VincentLanglet commented:
https://github.com/sonata-project/SonataPageBundle/runs/6769778295?check_suite_focus=true is failing

mareckigit commented:
@ValentinMerlet Consider to implement Service Subscriber in the BaseCommand and to implement [Lazy Loading](https://symfony.com/doc/5.4/console/commands_as_services.html#lazy-loading) in all commands. I'm working with my own [fork](https://github.com/treehousedevelopers/SonataPageBundle) that contains your changes and [Haniki:sf4_bump](https://github.com/Haniki/SonataPageBundle/tree/sf4_bump). Maybe some changes will be useful for you.

ValentinMerlet committed:
Use CmsManagerInterface instead of CmsPageManager because of final is planned

ValentinMerlet committed:
Add a CmsManagerServiceLocatorCompilerPass in order to add, as argument, to the
sonata.cms_manager.service_locator service, the services implementing CmsManagerInterface

ValentinMerlet committed:
Make Symfony commands lazy

ValentinMerlet committed:
Implement Service Subscriber in the BaseCommand

ValentinMerlet commented:
> @ValentinMerlet Consider to implement Service Subscriber in the BaseCommand and to implement [Lazy Loading](https://symfony.com/doc/5.4/console/commands_as_services.html#lazy-loading) in all commands. I'm working with my own [fork](https://github.com/treehousedevelopers/SonataPageBundle) that contains your changes and [Haniki:sf4_bump](https://github.com/Haniki/SonataPageBundle/tree/sf4_bump). Maybe some changes will be useful for you.

Done :)

VincentLanglet commented:
WDYT about the solution @jordisala1991 ? Is it something "classic" to implement service subscriber ? 
I was ok with the injection in the constructor IMHO

treehousedevelopers commented:
I have implemented Service Subscriber in my fork https://github.com/treehousedevelopers/SonataPageBundle/commit/066f3ea1dfecb401a9fb6123a9fc7c11ee2f2bf9

mareckigit commented:
I have implemented Service Subscriber in my fork treehousedevelopers@066f3ea

eerison commented on src/Command/BaseCommand.php:
```diff
@@ -13,78 +13,47 @@
-     * @return PageManagerInterface
-     */
-    public function getPageManager()
+    public function __construct(ContainerInterface $locator)
```
we need to avoid to inject the whole ContainerInterface, as far as I know it will be removed for the future symfony releases, and I guess to this work the service need to be public!

mareckigit replied:
This is not whole container. The interface is the same, but it is small container that includes only declared services and provides lazy. Symfony do the same job in controllers, You can read more about it in doc https://symfony.com/doc/5.4/service_container/service_subscribers_locators.html

eerison replied:
hmmm interesting, I didn't know about it!
well I need to research a bit more about this, But I would inject all dependencies explicit in my service!

But I do not have any argument to say that it's right or wrong, I just like to look at the class and see what is necessary for this work!

mareckigit replied:
I have suggested to use Service Subscriber because there is set of services here and commands use only some from the set. The benefit is that only used services are initialized not all from the set. Commands should be also lazy and private services.

eerison replied:
well but as it's supported by symfony 5.x and 6.x, I'm fine with that, in the future if someone could provide a better solution we can do :)

VincentLanglet replied:
I don't know this a lot too, so I pinged @jordisala1991 to give his opinion on this strategy.

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

VincentLanglet commented:
WDYT here @jordisala1991, should we inject services in the constructor or implements ServiceSubscriberInterface ?

jordisala1991 commented:
IMO inject deps on construct, and avoid a base class.

VincentLanglet commented:
Indeed the baseCommand is not useful at all, and one of the method `getNotificationBackend` is now deprecated

Without this base command, each command will have his own dependencies in their constructor.

eerison commented:
> 

I guess we won't need this baseCommand too, after refactor the cleanup command I can see what is really necessary or not!

eerison commented:
I checked the BaseCommand and as we thought, we won't need this class anymore, In my Pull request: https://github.com/sonata-project/SonataPageBundle/pull/1434/files#diff-a351a99354f8a5971094d057717392bb08fcfa1b8fa6310d2b8d54d219068552, I'm deprecating the baseCommand, then I guess this PR is not necessary anymore.

jordisala1991 commented:
This is already fixed on 4.x
